cranelift/codegen/meta/src/shared/formats.rs

```diff
@@ -206,6 +208,17 @@ impl Formats {
                 .imm_with_name("size", &imm.uimm8)
                 .build(),
 
+            heap_load: Builder::new("HeapLoad").imm(&imm.heap_imm).value().build(),
+
+            heap_store: Builder::new("HeapStore")
+                // We have more fields for this instruction than
+                // `InstructionData` can hold without growing in size, so we
+                // push the immediates out into a side table.
+                .imm(&imm.heap_imm)
+                .value()
+                .value()
+                .build(),
```

Comment:
So `heap_store` takes two `Value`s and a `Uimm32` (32 bits each), a `MemFlags` (8 bits), and a `Heap`, right? And we need two bytes in `InstructionData` for the opcode plus the enum discriminant. As a somewhat out-there suggestion: Can we make `Heap` be a `u8` underneath and avoid the side table? It would be surprising to have more than 256 separate heaps, right?

Reply:
It's pretty tempting, on the one hand... on the other hand, with component-model work, I suspect that eventually having multiple heaps accessible from one body of code (e.g. adapter functions) may be common, and 256 heaps starts to feel uncomfortably close to what may actually exist (e.g. O(100) heaps for O(100) modules in a large component graph)...

Reply:
(I guess in such a case one could argue that an adapter function should never need to see more than `from` and `to` heaps, rather than all heaps in the graph, so maybe it's OK. I'm not entirely sure, just trying to avoid creating surprising limits if we can)

Reply:
A related question: `Heap` is an index into the heaps declared on the current function, so this hypothetical 256-heap limit would be per-function; do I have that right? So a core wasm module, or a collection of components, could still have billions of heaps as long only a few are used in each function. I think that's assumed in your comment about adapter functions, but I don't understand the heap infrastructure well so I wanted to check.

I don't have strong feelings on this either way, it just seems like the instruction is so close to fitting without a side table that we ought to at least think about it a little.

Reply:
Indeed, it's per-function. I haven't looked closely at how cranelift-wasm actually defines heaps but I would suspect if it doesn't create descriptors on-demand as heaps are actually referenced, it wouldn't be hard to do so.

Reply:
Right now, all of our entities are `u32` so this feels like a larger change we would want to think over a bit. As such, I'm not going to block landing this PR on thinking about that.